### DISCUSSION TITLE
Tiny Editor Compiles Itself.

### DISCUSSION THREAD

**Discussion by Whammo** [jkotlinski/durexforth#437] [Show and tell]
It's not bullet-proof and doesn't account for graphics characters, but you can full screen edit one screen worth of code, and compile it with Shift-Return. A save to disk?

```forth
include io
: ssave ( addr -- )
19 emit 3 device 0 0 1 255 open ioabort
clrchn 1 chkin ioabort  1000 0 do chrin
readst if drop unloop then over i + c!
loop clrchn 1 close 8 device drop ;
: sogo $d6 c@ 24 = if drop else emit
then ;
: t page begin 0 $cc c! begin key?
until 1 $cc c! $cf c@ 0<> if 0 $cf
c! $ce c@ $d1 @ $d3 c@ + c! $287 c@ $f3
@ $d3 c@ + c! then key dup
$d6 c@ 24 = $d3 c@ 39 = and over dup
$91 <> swap $9d <> and and if 2drop 0
then case
0 of endof
$d of sogo endof
$11 of sogo endof
$8d of drop here $ff + dup ssave 1000
evaluate quit endof
emit endcase again ;
```

**Comment by Whammo** [comment]
It's not too difficult to keep the screen from scrolling but it limits screen to 999 bytes.

  ```forth
$d6 c@ 24 = $d3 c@ 39 = and over dup
$91 <> swap $9d <> and and if 2drop 0
```
 From a Facebook discussion reference,
 [https://www.lemon64.com/forum/viewtopic.php?p=880370&fbclid=IwAR2G_uAxWRz9VphIC9j_dNRpMyFbnCWLAKnf5lgTbieri2lNYJhLkvdCkwg#880370](url)

Documented:
```basic
999 rem Cursor Blink on
1000 poke204,0
1004 rem loop for char
1005 ifpeek(198)=0then1005
1009 rem blink off
1010 poke204,1
1014 rem Was Last Curson Blink reverse?
1015 ifpeek(207)=0then1040
1019 rem Blink to non reverse
1020 poke207,0
1023 rem poke the address of the current line + Cursor Column
1024 rem on Current Line with Character Under Cursor
1025 poke(peek(210)*256+peek(209)+peek(211)),peek(206)
1030 rem likewise to color ram
1030 poke(peek(244)*256+peek(243)+peek(211)),peek(647)
1040 geta$:printa$;:goto1000
```

**Reply by Whammo** [reply]
Issue: insert causes quote mode. shift-return is normally the get out of quote mode free card.
 Will also cause a scroll if on last line.
 Resolution: trap for insert, (inhibit?) quote mode. Use F-keys for major functions.
 Bonus: insert seems to handle long lines well, limits line to 80 characters.
Perhaps if I force a long line on second to last line, things will take care of itself. (wishful thinking?) 
Quote mode may be useful i.e. 'Q' emit

**Comment by ekipan** [comment]
To help myself understand this one too I rewrote parts of it and am now using it in a program of mine. I was able to make the keyboard interpreter rather simpler.

* Extracted several words `run x y`, sogo into `y24< ?emit`, cursor reset `unblink` as my program blinks the cursor in several places.
* Deleted the color memory code as I don't care about it.
* Removed the 0 case. `0 emit` doesn't seem to do anything. Let me know if I'm mistaken.
* Simplified the bottom check by special casing up and left to always emit and putting the check in the default case, reusing `y24<` and `?emit`.

In `ssave` is the phrase `if drop unloop then` intended to be `if drop leave then`? As is I think it crashes when there's a read error. I use blocks for saving and loading so my program doesn't use your ssave word.

```forth
include io
: ssave ( ... as before ... ) ;
: run  dup ssave 999 evaluate quit ;
: x  $d3 c@ ;    : ?emit  and emit ;
: y  $d6 c@ ;    : y24<  y 24 < ;
: unblink  1 $cc c! $cf c@ if 
0 $cf c! $ce c@ $d1 @ x + c! then ;
: t  page begin
0 $cc c! key dup unblink case
$91 of emit endof \ up
$9d of emit endof \ left
$11 of y24< ?emit endof \ down
$0d of y24< ?emit endof \ return
$8d of drop here $ff + run endof \ sret
x 39 < y24< or ?emit endcase again ;
```

To fix the insert-mode problem in my program I added the case

`$94 of emit space $9d emit endof`
 
But it's less than satisfactory.

**Reply by Whammo** [reply]
I was having fun doing it, but found puzzling out the screen editor was more work than starting from scratch. I don't want to start from scratch and believe one needs to understand the PET's screen editor before one can understand the 64's.

**Reply by Whammo** [reply]
If you notice, I open the screen as an output file in ```ssave``` .
This is the conversion to PETSCII from screen code.
 It's probably simple to pipe the screen output to the compiler, but I did not get that far.

**Comment by Whammo** [comment]
I believe I understand how to handle the screen editor without starting entirely from scratch.
The key, is in a custom IRQ handler.
